versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTxDatabaseAdapter.java

```diff
@@ -0,0 +1,751 @@
+                    targetBranch,
+                    expectedHead,
+                    targetHead,
+                    sequenceToTransplant,
+                    branchCommits,
+                    newKeyLists);
+
+            Hash newGlobalHead =
+                writeGlobalCommit(
+                    ctx, timeInMicros, Hash.of(pointer.getGlobalId()), Collections.emptyList());
+
+            return updateNamedRef(targetBranch, pointer, targetHead, newGlobalHead);
+
+            // 6. return hash of last commit added to 'targetBranch' (via the casOpLoop)
+          },
+          () ->
+              transplantConflictMessage(
+                  "Retry-failure", targetBranch, expectedHead, sequenceToTransplant));
+
+    } catch (ReferenceNotFoundException | ReferenceConflictException | RuntimeException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public Hash commit(CommitAttempt commitAttempt)
+      throws ReferenceConflictException, ReferenceNotFoundException {
+    try {
+      return casOpLoop(
+          commitAttempt.getCommitToBranch(),
+          false,
+          true,
+          (ctx, pointer, x, newKeyLists) -> {
+            Hash branchHead = branchHead(pointer, commitAttempt.getCommitToBranch());
+
+            long timeInMicros = commitTimeInMicros();
+
+            CommitLogEntry newBranchCommit =
+                commitAttempt(ctx, timeInMicros, branchHead, commitAttempt, newKeyLists);
+
+            Hash newGlobalHead =
+                writeGlobalCommit(
+                    ctx,
+                    timeInMicros,
+                    Hash.of(pointer.getGlobalId()),
+                    commitAttempt.getGlobal().entrySet().stream()
+                        .map(e -> ContentsIdAndBytes.of(e.getKey(), (byte) 0, e.getValue()))
+                        .collect(Collectors.toList()));
+
+            return updateNamedRef(
+                commitAttempt.getCommitToBranch(),
+                pointer,
+                newBranchCommit.getHash(),
+                newGlobalHead);
+          },
+          () ->
+              commitConflictMessage(
+                  "Retry-Failure",
+                  commitAttempt.getCommitToBranch(),
+                  commitAttempt.getExpectedHead()));
+    } catch (ReferenceNotFoundException | ReferenceConflictException | RuntimeException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public Hash create(NamedRef ref, Hash target)
+      throws ReferenceAlreadyExistsException, ReferenceNotFoundException {
+    try {
+      return casOpLoop(
+          ref,
+          false,
+          false,
+          (ctx, pointer, branchCommits, newKeyLists) -> {
+            if (pointer.getNamedReferencesMap().containsKey(ref.getName())) {
+              throw referenceAlreadyExists(ref);
+            }
+
+            Hash hash = target;
+            if (hash == null) {
```

Comment:
maybe: `Hash hash = null == target ? NO_ANCESTOR : target`

Reply:
Nah - I'd like to keep the comment ther.

Reply:
you can still keep the comment fwiw :)